server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java

```diff
@@ -35,13 +35,48 @@ public class ClusterInfoSimulator {
 
     public ClusterInfoSimulator(RoutingAllocation allocation) {
         this.allocation = allocation;
-        this.leastAvailableSpaceUsage = new HashMap<>(allocation.clusterInfo().getNodeLeastAvailableDiskUsages());
-        this.mostAvailableSpaceUsage = new HashMap<>(allocation.clusterInfo().getNodeMostAvailableDiskUsages());
+        this.leastAvailableSpaceUsage = getAdjustedDiskSpace(allocation, allocation.clusterInfo().getNodeLeastAvailableDiskUsages());
+        this.mostAvailableSpaceUsage = getAdjustedDiskSpace(allocation, allocation.clusterInfo().getNodeMostAvailableDiskUsages());
         this.shardSizes = new CopyOnFirstWriteMap<>(allocation.clusterInfo().shardSizes);
         this.shardDataSetSizes = Map.copyOf(allocation.clusterInfo().shardDataSetSizes);
         this.dataPath = Map.copyOf(allocation.clusterInfo().dataPath);
     }
 
+    /**
+     * Cluster info contains a reserved space that is necessary to finish initializing shards (that are currently in progress).
+     * for all initializing shards sum(expected size) = reserved space + already used space
+     * This deducts already used space from disk usage as when shard start is simulated it is going to add entire expected shard size.
+     */
+    private static Map<String, DiskUsage> getAdjustedDiskSpace(RoutingAllocation allocation, Map<String, DiskUsage> diskUsage) {
+        var diskUsageCopy = new HashMap<>(diskUsage);
```

Comment:
This might be modified further if `org.elasticsearch.cluster.ClusterInfoSimulator#modifyDiskUsage` is called with non-zero size.

It was always copied before, but let me check if I can use `CopyOnFirstWriteMap` here (with updates via iterators entries).

Reply:
This can be done with [this change](https://github.com/idegtiarenko/elasticsearch/commit/8fa1620c2c4c31d61750e9ed7019e6a5987e388d) at expense of additional `CopyOnFirstWriteEntry` allocated.

Reply:
Yeah I think that's probably worse than what we have here.